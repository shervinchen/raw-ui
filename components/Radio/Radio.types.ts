import { ChangeEvent, InputHTMLAttributes } from "react"

export type RadioValue = string | number | undefined

export interface BaseRadioProps {
  checked?: boolean
  defaultChecked?: boolean
  value?: RadioValue
  disabled?: boolean
  className?: string
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
}

type NativeRadioProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  keyof BaseRadioProps
>;

export type RadioProps = BaseRadioProps & NativeRadioProps
