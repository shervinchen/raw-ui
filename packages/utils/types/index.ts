export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
