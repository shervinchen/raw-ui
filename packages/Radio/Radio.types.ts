import { ChangeEvent, ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export type RadioValue = string | number | undefined;

export interface BaseRadioProps {
  checked?: boolean;
  defaultChecked?: boolean;
  value?: RadioValue;
  disabled?: boolean;
  className?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export type RadioProps = Merge<
  ComponentPropsWithoutRef<'input'>,
  BaseRadioProps
>;
