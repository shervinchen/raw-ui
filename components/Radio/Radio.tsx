import React, { FC, ChangeEvent, useEffect, useMemo } from "react";
import classNames from "classnames";
import { RadioProps } from "./Radio.types";
import { useControlled } from "../utils/hooks";
import { RawUITheme } from "../Theme/preset/preset.type";
import { useTheme } from "../Theme/theme-context";
import { useRadioGroupContext } from "./radio-group-context";

const Radio: FC<RadioProps> = ({
  defaultChecked = false,
  checked,
  value: radioValue,
  disabled = false,
  onChange,
  className = "",
  children,
  ...restProps
}) => {
  const theme: RawUITheme = useTheme();
  const { inGroup, groupDisabled, groupValue, onGroupChange } = useRadioGroupContext()
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue: defaultChecked,
    value: checked,
  });
  const isDisabled = inGroup ? groupDisabled || disabled : disabled;
  const selfChecked = useMemo(() => {
    if (!inGroup) {
      return internalValue
    }
    return groupValue === radioValue
  }, [internalValue, inGroup, groupValue, radioValue])

  const radioClasses = classNames("raw-radio", className);
  const radioInnerClasses = classNames(
    "raw-radio-inner",
    selfChecked && "checked"
  );

  const changeHandler = (event: ChangeEvent<HTMLInputElement>) => {
    if (isDisabled) return;
    setInternalValue(event.target.checked);
    if (inGroup) onGroupChange?.(radioValue)
    onChange?.(event);
  };

  return (
    <label className={radioClasses}>
      <input
        className="raw-radio-input"
        type="radio"
        disabled={isDisabled}
        checked={selfChecked}
        onChange={changeHandler}
        {...restProps}
      />
      <span className={radioInnerClasses} />
      {children && <span className="raw-radio-text">{children}</span>}
      <style jsx>
        {`
          .raw-radio {
            position: relative;
            display: inline-flex;
            align-items: center;
            cursor: ${isDisabled ? "not-allowed" : "pointer"};
            opacity: ${isDisabled ? 0.5 : 1};
            gap: 8px;
          }
          .raw-radio-input {
            position: absolute;
            opacity: 0;
            outline: none;
            width: 0;
            height: 0;
            margin: 0;
            padding: 0;
            z-index: -1;
            font-size: 0;
            background-color: transparent;
          }
          .raw-radio-inner {
            position: relative;
            width: 14px;
            height: 14px;
            border: 1px solid
              ${selfChecked ? theme.palette.black : theme.palette.accents7};
            border-radius: 50%;
          }
          .raw-radio-inner:after {
            content: "";
            display: block;
            width: 8px;
            height: 8px;
            border-radius: 50%;
            position: absolute;
            top: 50%;
            left: 50%;
            background-color: ${theme.palette.black};
            transition: transform 0.15s ease;
            transform: translate(-50%, -50%) scale(0);
          }
          .raw-radio-inner.checked:after {
            transform: translate(-50%, -50%) scale(1);
          }
          .raw-radio-text {
            font-size: 14px;
            line-height: normal;
            user-select: none;
          }
        `}
      </style>
    </label>
  );
};

export default Radio;
