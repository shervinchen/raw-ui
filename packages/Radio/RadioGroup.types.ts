import { ComponentPropsWithoutRef } from 'react';
import { RadioValue } from './Radio.types';
import { Merge } from '../utils';

export interface BaseRadioGroupProps {
  defaultValue?: RadioValue;
  value?: RadioValue;
  disabled?: boolean;
  layout?: 'row' | 'column';
  onChange?: (value: RadioValue) => void;
  className?: string;
}

export type RadioGroupProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseRadioGroupProps
>;

export interface RadioGroupConfig {
  groupDisabled: boolean;
  groupValue: RadioValue;
  inGroup: boolean;
  onGroupChange?: (radioValue?: RadioValue) => void;
}
