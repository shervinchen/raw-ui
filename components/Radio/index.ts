import Radio from "./Radio";
import RadioGroup from "./RadioGroup";

export type RadioComponentType = typeof Radio & {
  Group: typeof RadioGroup;
};

(Radio as RadioComponentType).Group = RadioGroup;

export default Radio as RadioComponentType;
