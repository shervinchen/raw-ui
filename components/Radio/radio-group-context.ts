import React, { createContext, useContext } from "react";
import { RadioGroupConfig } from "./RadioGroup.types";

const defaultContext: RadioGroupConfig = {
  groupDisabled: false,
  groupValue: '',
  inGroup: false,
};

export const RadioGroupContext =
  createContext<RadioGroupConfig>(defaultContext);

export const useRadioGroupContext = (): RadioGroupConfig =>
  useContext<RadioGroupConfig>(RadioGroupContext);
