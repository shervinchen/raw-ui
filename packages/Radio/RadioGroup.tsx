import React, { FC, PropsWithChildren, useCallback, useMemo } from 'react';
import classNames from 'classnames';
import { RadioGroupConfig, RadioGroupProps } from './RadioGroup.types';
import { useControlled } from '../utils/hooks';
import { RadioValue } from './Radio.types';
import { RadioGroupContext } from './radio-group-context';

const RadioGroup: FC<PropsWithChildren<RadioGroupProps>> = ({
  value: radioGroupValue,
  defaultValue = '',
  disabled = false,
  className = '',
  layout = 'row',
  onChange,
  children,
  ...restProps
}) => {
  const [internalValue, setInternalValue] = useControlled<RadioValue>({
    defaultValue,
    value: radioGroupValue,
  });
  const classes = classNames('raw-radio-group', className);

  const handleGroupChange = useCallback(
    (radioValue?: RadioValue) => {
      setInternalValue(radioValue);
      onChange?.(radioValue);
    },
    [onChange, setInternalValue]
  );

  const radioGroupConfig = useMemo<RadioGroupConfig>(() => {
    return {
      groupValue: internalValue,
      groupDisabled: disabled,
      inGroup: true,
      onGroupChange: handleGroupChange,
    };
  }, [internalValue, disabled, handleGroupChange]);

  return (
    <RadioGroupContext.Provider value={radioGroupConfig}>
      <div role="radiogroup" className={classes} {...restProps}>
        {children}
        <style jsx>{`
          .raw-radio-group {
            display: inline-flex;
            flex-direction: ${layout};
            gap: 32px;
          }
        `}</style>
      </div>
    </RadioGroupContext.Provider>
  );
};

export default RadioGroup;
