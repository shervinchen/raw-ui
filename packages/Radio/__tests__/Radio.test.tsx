import React, { ChangeEvent, useState } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Radio from '..';
import { RadioProps } from '../Radio.types';

const optionsData = [
  {
    name: 'React',
    value: 'react',
  },
  {
    name: 'Vue',
    value: 'vue',
  },
  {
    name: 'Angular',
    value: 'angular',
  },
  {
    name: 'Svelte',
    value: 'svelte',
  },
];

describe('Radio', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Radio />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    const { container } = render(<Radio className="custom-radio" />);
    expect(container.firstChild).toHaveClass('custom-radio');
  });

  test('should trigger event when clicked', async () => {
    const clickHandler = jest.fn();
    const { container } = render(<Radio onClick={clickHandler} />);
    const radio = container.firstChild;
    await userEvent.click(radio as Element);
    expect(clickHandler).toHaveBeenCalledTimes(1);
  });

  test('should support uncontrolled value', () => {
    const { container } = render(<Radio defaultChecked />);
    const radioInput = container.querySelector('.raw-radio-input');
    expect(radioInput).toBeChecked();
  });

  test('should support label text', () => {
    const { queryByText } = render(<Radio>Label</Radio>);
    expect(queryByText('Label')).toBeInTheDocument();
  });

  test('should support disabled', async () => {
    const { container } = render(<Radio disabled />);
    const radio = container.firstChild;
    const radioInput = container.querySelector('.raw-radio-input');
    expect(radioInput).toBeDisabled();
    await userEvent.click(radio as Element);
    expect(radioInput).not.toBeChecked();
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: RadioProps) => {
      const [checked, setChecked] = useState(false);

      return (
        <Radio
          checked={checked}
          onChange={(event: ChangeEvent<HTMLInputElement>) => {
            setChecked(event.target.checked);
            props.onChange?.(event);
          }}
        >
          Controlled
        </Radio>
      );
    };

    const { container } = render(<Component onChange={onChange} />);
    const radio = container.firstChild;
    const radioInput = container.querySelector('.raw-radio-input');
    expect(radioInput).not.toBeChecked();
    await userEvent.click(radio as Element);
    expect(radioInput).toBeChecked();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('should support radio group uncontrolled value', async () => {
    const Component = () => (
      <Radio.Group defaultValue="react">
        {optionsData.map((item) => (
          <Radio value={item.value} key={item.value}>
            {item.name}
          </Radio>
        ))}
      </Radio.Group>
    );
    const { container } = render(<Component />);
    const radioOne = container.querySelectorAll('input')[0];
    const radioTwo = container.querySelectorAll('input')[1];
    const radioThree = container.querySelectorAll('input')[2];
    const radioFour = container.querySelectorAll('input')[3];
    expect(radioOne).toBeChecked();
    expect(radioTwo).not.toBeChecked();
    expect(radioThree).not.toBeChecked();
    expect(radioFour).not.toBeChecked();
    await userEvent.click(radioTwo);
    expect(radioOne).not.toBeChecked();
    expect(radioTwo).toBeChecked();
    expect(radioThree).not.toBeChecked();
    expect(radioFour).not.toBeChecked();
  });

  test('should support radio group controlled value', async () => {
    let checked = 'react';

    const onChange = jest.fn((value: string) => {
      checked = value;
    });

    const Component = (props) => (
      <Radio.Group {...props}>
        {optionsData.map((item) => (
          <Radio value={item.value} key={item.value}>
            {item.name}
          </Radio>
        ))}
      </Radio.Group>
    );

    const { container, rerender } = render(
      <Component value={checked} onChange={onChange} />
    );

    const [radioOne, radioTwo, radioThree, radioFour] = Array.from(
      container.querySelectorAll('input')
    );

    expect(radioOne).toBeChecked();
    expect(radioTwo).not.toBeChecked();
    expect(radioThree).not.toBeChecked();
    expect(radioFour).not.toBeChecked();

    await userEvent.click(radioTwo);
    rerender(<Component value={checked} onChange={onChange} />);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(checked).toEqual('vue');
  });

  test('should support radio group disabled', async () => {
    const Component = () => (
      <Radio.Group defaultValue="" disabled>
        <Radio value="react">React</Radio>
        <Radio value="vue">Vue</Radio>
        <Radio value="angular" disabled>
          Angular
        </Radio>
        <Radio value="svelte" disabled={false}>
          Svelte
        </Radio>
      </Radio.Group>
    );
    const { container } = render(<Component />);
    const [radioOne, radioTwo, radioThree, radioFour] = Array.from(
      container.querySelectorAll('input')
    );
    expect(radioOne).toBeDisabled();
    expect(radioTwo).toBeDisabled();
    expect(radioThree).toBeDisabled();
    expect(radioFour).toBeDisabled();

    await userEvent.click(radioOne);
    await userEvent.click(radioTwo);
    await userEvent.click(radioThree);
    await userEvent.click(radioFour);

    expect(radioOne).not.toBeChecked();
    expect(radioTwo).not.toBeChecked();
    expect(radioThree).not.toBeChecked();
    expect(radioFour).not.toBeChecked();
  });

  test('should support row and col layout in radio group', () => {
    const Component = ({ layout }: { layout: 'row' | 'column' }) => (
      <Radio.Group defaultValue="react" layout={layout}>
        {optionsData.map((item) => (
          <Radio value={item.value} key={item.value}>
            {item.name}
          </Radio>
        ))}
      </Radio.Group>
    );
    const { container, rerender } = render(<Component layout="row" />);
    const radioGroup = container.firstChild as Element;
    expect(getComputedStyle(radioGroup).flexDirection).toBe('row');
    rerender(<Component layout="column" />);
    expect(getComputedStyle(radioGroup).flexDirection).toBe('column');
  });
});
