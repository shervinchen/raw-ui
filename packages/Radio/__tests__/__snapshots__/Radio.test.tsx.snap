// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should match the snapshot 1`] = `
<DocumentFragment>
  <label
    class="jsx-2317508149 raw-radio"
  >
    <input
      class="jsx-1802955700 raw-radio-input"
      type="radio"
    />
    <span
      class="jsx-2317508149 raw-radio-inner"
    />
  </label>
</DocumentFragment>
`;
